extensions/amp-analytics/0.1/test/test-linker.js

```diff
@@ -67,40 +79,69 @@ describe('Linker', () => {
         color: 'green',
         car: 'tesla',
       },
-      output: '1~1eit8v3~key1~value1~name~bob~color~green~car~tesla',
+      output: '1*efercj*key1*dmFsdWUx*name*Ym9i*color*Z3JlZW4.*car*dGVzbGE.',
     },
     {
-      description: 'fake macros',
+      description: 'value contains URL unsafe chars',
       version: '1',
       pairs: {
         cid: '12345',
         ref: 'https://www.example.com',
       },
-      output: '1~s21iu~cid~12345~ref~https%3A%2F%2Fwww.example.com',
+      output: '1*vz1vnu*cid*MTIzNDU.*ref*aHR0cHM6Ly93d3cuZXhhbXBsZS5jb20.',
     },
     {
-      description: 'encodes url safe keys and values',
+      description: 'works if value has *',
       version: '1',
       pairs: {
-        '<unsafe>': '//foo@bar',
+        'key': '*hi*',
+        'key2': '***',
       },
-      output: '1~1gbjnst~%3Cunsafe%3E~%2F%2Ffoo%40bar',
+      output: '1*lpias9*key*KmhpKg..*key2*Kioq',
     },
     {
-      description: 'encodes tilde in key value pairs',
+      description: 'encodes unicode in values',
       version: '1',
       pairs: {
-        '~key~': 'hi',
-        'key2': '~hi~',
+        'key': '中文',
+        'key2': 'π',
       },
-      output: '1~1wqggkq~%7Ekey%7E~hi~key2~%7Ehi%7E',
+      output: '1*1pf8zsw*key*5Lit5paH*key2*z4A.',
     },
   ];
 
   tests.forEach(test => {
     it(test.description, () => {
-      const result = createLinker(test.version, test.pairs);
-      return expect(result).to.equal(test.output);
+      expect(createLinker(test.version, test.pairs)).to.equal(test.output);
+    });
+  });
+
+  const testsWithInvalidKeys = [
+    {
+      description: 'ignore invalid keys',
+      version: '1',
+      pairs: {
+        '*invalid': '123',
+        'valid': 'abc',
+      },
+      output: '1*bn7mun*valid*YWJj',
+    },
+    {
+      description: 'return empty string if all keys are invalid',
+      version: '1',
+      pairs: {
+        '*invalid': '123',
+        'invalid!': 'abc',
+      },
+      output: '',
+    },
+  ];
+
+  testsWithInvalidKeys.forEach(test => {
+    it(test.description, () => {
+      allowConsoleError(() => {
+        expect(createLinker(test.version, test.pairs)).to.equal(test.output);
+      });
     });
```

Comment:
Could you please add a test that makes sure that AMP linker will be understood by GA parser

I've created a sample vector:
 const clientIds = [
      {
        clientId: 'GA1.3.1218435055.1536188913',
        encoded: '1*itvlvc*_ga*R0ExLjMuMTIxODQzNTA1NS4xNTM2MTg4OTEz',
      },

      {
        clientId: 'GA1.2.amp-' +
            'oRg8vByriPdstwLgkz-UNWbp2P13vNFsnhES5vW8s5WodTOoea0mTiY7X62utLyz',
        encoded: '1*of6055*_ga*R0ExLjIuYW1wLW9SZzh2QnlyaVBkc3R3TGdrei1VTldicDJQMTN2TkZzbmhFUzV2VzhzNVdvZFRPb2VhMG1UaVk3WDYydXRMeXo.',
      },
      {
        clientId:
            'WgcaAD4XN2lydhQVNFruk6X8zwoUg6K2RnaRlhjs6CXvTv4aJV-3oVLdI1WxxvJb',
        encoded: '1*19eaxqc*_ga*V2djYUFENFhOMmx5ZGhRVk5GcnVrNlg4endvVWc2SzJSbmFSbGhqczZDWHZUdjRhSlYtM29WTGRJMVd4eHZKYg..',
      },
    ];

(assuming linker encodes pair  _ga:clietnId)

Reply:
added those test cases.
note that 
1) the checksum will be all different. 
2) Prefix `GA1.3.` will not be forwarded since it's not part of CLIENT_ID. Receiver needs to add them when saving to cookie

To get the exact same checksum, you could mock the clock and fingerprint as the following:
```js
    sandbox.stub(Date.prototype, 'getTimezoneOffset').returns(400);
    const mockWin = mockWindowInterface(sandbox);
    mockWin.getUserAgent.returns('Chrome 70');
    mockWin.getUserLanguage.returns('en-US');
```

Reply:
Yeah, my bad, I updated the existing test and it already had necessary data mocked

   currentTime = 1533329483292;
    timezone = 420;
    uaString = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, ' +
        'like Gecko) Chrome/67.0.3396.99 Safari/537.36';
    language = 'en-US';

Reply:
It looks like you've already merged the request, let me update tests on our side then

Reply:
BTW, did you mock the timestamp? cannot find it (should affect checksum)

Reply:
Yes, time is stubbed here https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/test/test-linker.js#L26